prusti-specs/src/lib.rs

```diff
@@ -128,3 +128,150 @@ pub fn body_invariant(tokens: TokenStream) -> TokenStream {
         }
     }
 }
+
+/** Unlike the functions above, which are only called from
+ *  prusti-contracts-internal, this function also needs to be called
+ *  from prusti-contracts-impl, because we still need to parse the
+ *  macro in order to replace it with the closure definition.
+ *  Therefore, there is an extra parameter drop_spec here which tells
+ *  the function whether to keep the specification (for -internal) or
+ *  drop it (for -impl).
+ */
+pub fn closure(tokens: TokenStream, drop_spec: bool) -> TokenStream {
+    let mut requires: Vec<TokenTree> = vec! [];
+    let mut ensures: Vec<TokenTree> = vec! [];
+    let mut cl: Option<syn::ExprClosure> = None;
+
+    let mut iter = tokens.clone ().into_iter();
+
+    while let Some(tok) = iter.next() {
```

Comment:
The parsing code should be either in `prusti-dev/prusti-specs/src/specifications/preparser.rs` or in `prusti-dev/prusti-specs/src/specifications/untyped.rs`. Also, is there any reason why you are using the tokens directly and not the higher level parsing API provided by the `syn` crate?

Reply:
But I'm not parsing a specification (well, sort of, but not in the sense of `preparser.rs`), I'm parsing the macro _containing_ the specification, so I think putting it into `prusti-specs/src/specifications/` would not be appropriate.

And I _am_ using the higher-level API when parsing the closure, the rest is just reading in the pre- and postconditions.

Reply:
> I think putting it into `prusti-specs/src/specifications/` would not be appropriate.

You could also put it in a separate file. If I am not mistaken, this function is the only one that does parsing in this file and, therefore, does not really fit.

> And I am using the higher-level API when parsing the closure, the rest is just reading in the pre- and postconditions.

I would still suggest defining something like:
```rust
struct ClosureWithSpecs {
  pres: ...,
  posts: ...,
  closure: ...,
}
```
and implementing `Parse` for it because it results in significantly more readable and maintainable code.

Reply:
Fixed.